modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1403 @@
+  if err_out != 0 {
+    throw SystemError.fromSyserr(err_out, "Failed to close tcpListener");
+  }
+  this.socketFd = -1;
+}
+
+/*
+  Returns the address on which socket is
+  listening on and bound to.
+
+  :return: bound address
+  :rtype: `ipAddr`
+*/
+proc tcpListener.addr throws {
+  return getsockname(this.socketFd);
+}
+
+inline operator !=(const ref lhs: tcpListener,const ref rhs: tcpListener) {
+  return lhs.socketFd != rhs.socketFd;
+}
+
+inline operator ==(const ref lhs: tcpListener,const ref rhs: tcpListener) {
+  return !(lhs.socketFd != rhs.socketFd);
+}
+
+proc tcpListener.writeThis(f) throws {
+  f.write("(","addr:",this.addr,",fd:",this.socketFd);
+}
+
+pragma "no doc"
+extern const SOMAXCONN: int;
+/*
+  Default `backlog` value used in :proc:`listen`
+  It is calulated as min(`SOMAXCONN`, 128) where `SOMAXCONN` is
+  the maximum number of allowed pending connections in the system.
+*/
+var backlogDefault:uint(16) = (if SOMAXCONN <= 128 then SOMAXCONN else 128):uint(16);
+
+/*
+  Convenience procedure which creates a new :type:`tcpListener` bound
+  to and listening on `address` for new connections. `backlog`
+  determines how many connections can be pending (not having called
+  accept) before the socket will begin to reject them. The default
+  value of backlog is `backlogDefault`.
+
+  .. code-block:: Chapel
+
+    const address = ipAddr.create("127.0.0.1", 8000, IPFamily.IPv4);
+    const server = listen(address);
+
+  :arg address: address to connect to
+  :type address: :type:`ipAddr`
+  :arg timeout: standard ipv6 address.
+  :type timeval: :type:`~Sys.timeval`
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: On failure to bind or listen on `address`
+*/
+proc listen(in address: ipAddr, reuseAddr: bool = true, backlog: uint(16) = backlogDefault): tcpListener throws {
+  var family = address.family;
+  var socketFd = socket(family, SOCK_STREAM);
+  bind(socketFd, address, reuseAddr);
+  var err_out = sys_listen(socketFd, backlog:c_int);
+  if err_out != 0 {
+    throw SystemError.fromSyserr(err_out, "Failed to listen on socket");
+  }
+  const listener = new tcpListener(socketFd);
+  return listener;
+}
+
+/*
+  Convenience procedure which creates a :type:`tcpConn` connected to
+  `address`.`timeout` determines how much time to wait for
+  connection to be established. The default value for `timeout` is
+  indefinite.
+
+  .. code-block:: Chapel
+
+    const address = ipAddr.create("127.0.0.1", 8000, IPFamily.IPv4);
+    const timeout = new timeval(4,0);
+    const connectedClient = connect(address, timeout);
+
+  :arg address: address to connect to
+  :type address: :type:`ipAddr`
+  :arg timeout: time to wait for connection establishment.
+  :type timeval: :type:`~Sys.timeval`
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: Upon failure to connect.
+*/
+proc connect(const ref address: ipAddr, in timeout = new timeval(-1,0)): tcpConn throws {
```

Comment:
Should we allow `real`s as timeouts?

Reply:
I think a structure holding these values is a bit more intuitive otherwise both at user and library level we need to do trivial tranformation for getting `seconds` and `microseconds` out of the real value.

Reply:
I didn't do a lot of socket programming but:

> I think a structure holding these values is a bit more intuitive

I disagree with this very strongly :)

Reply:
@e-kayrakli - in an offline conversation I had with @king-11 it's my understanding that he's going to look at providing support for both these structures as well as `real` to represent time in seconds.

Reply:
In case I wasn't clear; I have no problem with the current signature, especially if it is following some sort of precedence and/or have other motivations. I was just suggesting of adding a `real` version alongside this one (which what you're saying).